### PR TITLE
Add SFTP public key authentication support

### PR DESCRIPTION
**Pull request by rnicoll**
Adds a new SFTP external file plugin which uses public key based authentication instead of passwords.

A number of private fields/functions in the existing SFTP class have been made protected in order to make extension of that class feasible, as part of this work.

When adding a new external file store, the first field is now automatically selected. This is a minor UI improvement, but also caught by the SFTP key UI in order to trigger key pair generation.

### CODE REVIEW

File: apps/files_external/lib/sftp_key.php [hunk 1]
```diff
@@ -0,0 +1,148 @@
...
+
+/**
+* Uses phpseclib's Net_SFTP class and the Net_SFTP_Stream stream wrapper to
+* provide access to SFTP servers.
+*/
+class SFTP_Key extends \OC\Files\Storage\SFTP {
+	private $public_key;
+	private $private_key;
+
+	public function __construct($params) {
```
Comment (rnicoll) [inline apps/files_external/lib/sftp_key.php]: For my reference; attempting to enforce the type ends with having to modify the signature of the abstract class "Common" in lib/private/files/storage/common.php

**Comment by LukasReschke** [comment]
@owncloud-bot Retest this please.

File: apps/files_external/lib/sftp_key.php [hunk 2]
```diff
@@ -0,0 +1,148 @@
+<?php
+/**
+ * Copyright (c) 2014 University of Edinburgh <Ross.Nicoll@ed.ac.uk>
+ * This file is licensed under the Affero General Public License version 3 or
+ * later.
+ * See the COPYING-README file.
+ */
+namespace OC\Files\Storage;
```
Comment (rnicoll) [inline apps/files_external/lib/sftp_key.php]: Ah... changing this has a huge knock-on impact in namespaces elsewhere. Certainly changing the namespace on the existing sftp.php causes errors, and changing just sftp_key.php cause the configuration to fail to save/load. Probably a good idea for someone to look at namespaces used throughout files_external, but it's a significantly wider scope of change than I'm trying to tackle here.

File: apps/files_external/appinfo/app.php [hunk 3]
```diff
@@ -177,5 +178,17 @@
...
 
+OC_Mount_Config::registerBackend('\OC\Files\Storage\SFTP_Key', array(
+	'backend' => 'SFTP with secret key login',
+	'priority' => 100,
+	'configuration' => array(
+		'host' => (string)$l->t('Host'),
+		'user' => (string)$l->t('Username'),
+		'public_key' => (string)$l->t('Public key'),
+		'private_key' => '#private_key',
+		'root' => '&'.$l->t('Root')),
```
Comment (rnicoll) [inline apps/files_external/appinfo/app.php]: More or less a direct cut & paste from the SFTP plugin definition above. Will test and find out what type is coming back.

File: apps/files_external/appinfo/app.php [hunk 4]
```diff
@@ -177,5 +178,17 @@
 		'password' => '*'.$l->t('Password'),
 		'root' => '&'.$l->t('Remote subfolder'))));
 
+OC_Mount_Config::registerBackend('\OC\Files\Storage\SFTP_Key', array(
```
Comment (rnicoll) [inline apps/files_external/appinfo/app.php]: Matches style used elsewhere in the same file. Happy to change them all if that's desirable?

File: apps/files_external/appinfo/app.php (hunk 3 above)
Comment (mmattel) [inline apps/files_external/appinfo/app.php]: Please change:
 ``'root' => '&'.$l->t('Root'))`` --> ``'root' => '&'.$l->t('Remote subfolder'))``
as we just have harmonized the terms used recently

Please do not forget to document this in:
https://github.com/owncloud/documentation/blob/master/admin_manual/configuration/external_storage_configuration.rst
https://github.com/owncloud/documentation/blob/master/admin_manual/configuration/external_storage_configuration_gui.rst

File: apps/files_external/ajax/sftp_key.php [hunk 5]
```diff
@@ -0,0 +1,21 @@
+<?php
+set_include_path(get_include_path().PATH_SEPARATOR.
+	\OC_App::getAppPath('files_external').'/3rdparty/phpseclib/phpseclib/phpseclib');
```
Comment (rnicoll) [inline apps/files_external/ajax/sftp_key.php]: Copy & paste again; finds the files fine without, removed (and now matches style in the SFTP plugin)

Comment (LukasReschke) [inline apps/files_external/ajax/sftp_key.php]: Yeah. Some of the files_external files are really old and you better don't copy paste from them :smile:

File: apps/files_external/controller/ajaxcontroller.php [hunk 6]
```diff
@@ -0,0 +1,47 @@
...
+	/**
+	 * Generates an SSH public/private key pair.
+	 *
+	 * @NoAdminRequired
+	 * @NoCSRFRequired
+	 */
+	public function getSshKeys() {
+		$key = this::generateSshKeys();
+
+		OCP\JSON::success(array('data' => array(
```
Comment (LukasReschke) [inline apps/files_external/controller/ajaxcontroller.php]: That would be a JSONResponse()

File: apps/files_external/controller/ajaxcontroller.php [hunk 7]
```diff
@@ -0,0 +1,47 @@
...
+	}
+
+	/**
+	 * Generates an SSH public/private key pair.
+	 *
+	 * @NoAdminRequired
+	 * @NoCSRFRequired
+	 */
+	public function getSshKeys() {
+		$key = this::generateSshKeys();
```
Comment (LukasReschke) [inline apps/files_external/controller/ajaxcontroller.php]: That would be $this->

File: apps/files_external/controller/ajaxcontroller.php [hunk 8]
```diff
@@ -0,0 +1,47 @@
...
+                $key = $rsa->createKey();
+                // Replace the placeholder label with a more meaningful one
+                $key['publicKey'] = str_replace('phpseclib-generated-key', gethostname(), $key['publickey']);
+	}
+
+	/**
+	 * Generates an SSH public/private key pair.
+	 *
+	 * @NoAdminRequired
+	 * @NoCSRFRequired
```
Comment (LukasReschke) [inline apps/files_external/controller/ajaxcontroller.php]: Remove that - that is a essential security check

File: apps/files_external/controller/ajaxcontroller.php [hunk 9]
```diff
@@ -0,0 +1,47 @@
...
+use OC\Preview;
+
+class AjaxController extends Controller {
+	public function __construct($appName, IRequest $request) {
+		parent::__construct($appName, $request);
+	}
+
+	private static function generateSshKeys() {
+		$rsa = new Crypt_RSA();
+                $rsa->setPublicKeyFormat(CRYPT_RSA_PUBLIC_FORMAT_OPENSSH);
```
Comment (LukasReschke) [inline apps/files_external/controller/ajaxcontroller.php]: Tabs instead spaces

File: apps/files_external/controller/ajaxcontroller.php [hunk 10]
```diff
@@ -0,0 +1,47 @@
...
+ * See the COPYING-README file.
+ */
+
+namespace OCA\Files_External\Controller;
+
+use OCP\AppFramework\Http;
+use OCP\AppFramework\Controller;
+use OCP\IRequest;
+use OCP\AppFramework\Http\JSONResponse;
+use OC\Preview;
```
Comment (LukasReschke) [inline apps/files_external/controller/ajaxcontroller.php]: Unused imports

**Comment by rnicoll** [comment]
Have to hop onto another project for the rest of the day, but that's a lot of the cleanup done. Needs tests still, and I need to come back to the redesign around AppFramework. Trying to merge in some AppFramework within an existing application (files_external) which is layered on another app (files) isn't as easy as it looked initially, but should be possible.

On Transifex; it seems to complain bitterly if I don't put in that string. I presume the need for a new translation will be picked up by the automated tools and it will, in time, "just work"? Haven't actually worked with Transifex before, I'll admit...

File: apps/files_external/ajax/sftp_key.php [hunk 11]
```diff
@@ -0,0 +1,18 @@
...
+$rsa->setPassword(\OC::$server->getConfig()->getSystemValue('secret', ''));
+
+$key = $rsa->createKey();
+// Replace the placeholder label with a more meaningful one
+$key['publickey'] = str_replace('phpseclib-generated-key', gethostname(), $key['publickey']);
+
+OCP\JSON::success(array('data' => array(
+	'private_key' => $key['privatekey'],
+	'public_key' => $key['publickey']
+)));
```
Comment (LukasReschke) [inline apps/files_external/ajax/sftp_key.php]: Should also be in a controller

File: apps/files_external/controller/ajaxcontroller.php (hunk 6 above)
Comment (rnicoll) [inline apps/files_external/controller/ajaxcontroller.php]: Oh, crud, that file was meant to be shoved in a branch before I pushed, please ignore. Hangon, I'll take it out again...

File: apps/files_external/controller/ajaxcontroller.php [hunk 12]
```diff
@@ -0,0 +1,47 @@
...
+
+class AjaxController extends Controller {
+	public function __construct($appName, IRequest $request) {
+		parent::__construct($appName, $request);
+	}
+
+	private static function generateSshKeys() {
+		$rsa = new Crypt_RSA();
+                $rsa->setPublicKeyFormat(CRYPT_RSA_PUBLIC_FORMAT_OPENSSH);
+                $rsa->setPassword(\OC::$server->getConfig()->getSystemValue('secret', ''));
```
Comment (LukasReschke) [inline apps/files_external/controller/ajaxcontroller.php]: You can inject that. With ownCloud 8 you should even be able to use automatic dependency resolution.

For that you can remove `application.php` again and then in this controller you change:

1. $appName to $AppName
2. Add `IConfig $config` and assign it to a class member

File: apps/files_external/controller/ajaxcontroller.php (hunk 6 above)
Comment (rnicoll) [inline apps/files_external/controller/ajaxcontroller.php]: Gone, sorry about that, is just a work in progress (hence it's a total mess)

File: apps/files_external/lib/sftp_key.php (hunk 2 above)
Comment (Xenopathic) [inline apps/files_external/lib/sftp_key.php]: Leave it as it is - files_external is a mess. It'll be dealt with at some point :smile:

**Comment by LukasReschke** [comment]
> On Transifex; it seems to complain bitterly if I don't put in that string. I presume the need for a new translation will be picked up by the automated tools and it will, in time, "just work"? Haven't actually worked with Transifex before, I'll admit...

The process is the following:

1. In the PHP file you use the English string and the l10n functions
2. A bot will once per day scan the strings in the PHP files and upload them to transifex
3. Users will translate the string on Transifex
4. The bot will fetch the translations and upload them to this repository

No guarantees on 100% correctness, something like that should it be ;-9

File: apps/files_external/lib/sftp_key.php [hunk 13]
```diff
@@ -0,0 +1,168 @@
...
+	/**
+	 * Throws an exception if the provided port number is invalid (cannot be resolved
+         * and is not an IPv4 address).
+         */
+	protected static function assertPortNumberValid($port) {
+		if (!preg_match('/^\d+$/', $port)) {
+			throw new \Exception('Port number must be a number.');
+		}
+		if ($port < 0
+			|| $port > 65535) {
```
Comment (Xenopathic) [inline apps/files_external/lib/sftp_key.php]: Is it necessary for these to be on separate lines?

File: apps/files_external/lib/sftp_key.php [hunk 14]
```diff
@@ -0,0 +1,168 @@
...
+			return null;
+		}
+		return $key;
+	}
+
+	/**
+	 * Throws an exception if the provided host name/address is invalid (cannot be resolved
+	 * and is not an IPv4 address).
+	 */
+	protected static function assertHostAddressValid($hostname) {
```
Comment (LukasReschke) [inline apps/files_external/lib/sftp_key.php]: No need for the static keyword.

File: apps/files_external/lib/sftp_key.php [hunk 15]
```diff
@@ -0,0 +1,168 @@
...
+			// Hostname is not an IPv4 address and cannot be resolved via DNS
+			throw new \Exception('Cannot resolve hostname.');
+		}
+	}
+
+	/**
+	 * Throws an exception if the provided port number is invalid (cannot be resolved
+         * and is not an IPv4 address).
+         */
+	protected static function assertPortNumberValid($port) {
```
Comment (LukasReschke) [inline apps/files_external/lib/sftp_key.php]: No need for the static keyword.

File: apps/files_external/appinfo/app.php (hunk 4 above)
Comment (LukasReschke) [inline apps/files_external/appinfo/app.php]: I'd go with just this one. - Others can be changed when we touch them.

File: apps/files_external/js/settings.js [hunk 16]
```diff
@@ -354,6 +354,7 @@ $(document).ready(function() {
 				return false;
 			}
 		});
+		$td.children('input.added').first().focus();
```
Comment (Xenopathic) [inline apps/files_external/js/settings.js]: This is unneeded - look 4 lines above: `$td.children().not('[type=hidden]').first().focus();`. Unless this does something the existing one didn't?

Comment (rnicoll) [inline apps/files_external/js/settings.js]: Nope, that's a duplicate from merging commits initially targeting v7. Thanks for spotting, removed.

File: apps/files_external/appinfo/app.php (hunk 3 above)
Comment (Xenopathic) [inline apps/files_external/appinfo/app.php]: The `(string)` thing was introduced by @wakeup - perhaps he could shed some light on it?

**Comment by rnicoll** [comment]
Still not done with the cleanup, just pushing updates so you can see progress. Should be much closer now though.

File: apps/files_external/appinfo/app.php [hunk 17]
```diff
@@ -177,5 +178,17 @@
...
+	'backend' => 'SFTP with secret key login',
+	'priority' => 100,
+	'configuration' => array(
+		'host' => (string)$l->t('Host'),
+		'user' => (string)$l->t('Username'),
+		'public_key' => (string)$l->t('Public key'),
```
Comment (wakeup) [inline apps/files_external/appinfo/app.php]: FYI: @rnicoll @Xenopathic casting to `string` was necessary to make the strings visible. This is the correct way though. If there are other chars like `*` or `&` no casting is necessary.

File: apps/files_external/lib/sftp.php [hunk 18]
```diff
@@ -42,7 +42,11 @@ public function __construct($params) {
 			$this->host = substr($this->host, $proto+3);
 		}
 		$this->user = $params['user'];
-		$this->password = $params['password'];
+		if (array_key_exists('password', $params)) {
+			$this->password = $params['password'];
+		} else {
+			$this->password = '';
+		}
```
Comment (Xenopathic) [inline apps/files_external/lib/sftp.php]: I'd suggest using the ternary operator form here as used elsewhere (see `root` right below)

**Comment by rnicoll** [comment]
Squashed the whole lot down, that should be everything but the tests done now.

**Comment by rnicoll** [comment]
Barring two tests (testHashInFileName and testRenameOverWriteDirectory) which also fail with the existing SFTP plugin (and I think are therefore more about how my server works, than anything else), all unit tests pass. Any other changes required?

**Comment by Xenopathic** [comment]
Looks good! I'll be testing this at some point, then I can give my review mark :smile:

**Comment by LukasReschke** [comment]
@karlitschek Please verify https://github.com/owncloud/core/pull/13190#issuecomment-69321430. - THX.

**Comment by karlitschek** [comment]
@LukasReschke Confirmed :-)

**Comment by Xenopathic** [comment]
Neat! I like the auto-generation of SSH keys, it makes it really simple to use. Only issue is a minor problem with the position of the 'Generate keys' button: it drops underneath the divider line and looks a bit odd:

screen

It seems that the button is an `<a>` element, while it should be an `<input>`, as it then gets the CSS necessary to put it inline with the other elements.

Other than that though, it's looking great!

**Comment by rnicoll** [comment]
Done. I'll update Google and Dropbox plugins to match in a moment.

**Comment by Xenopathic** [comment]
Nice! :+1: Now we wait for feature merges to be accepted again

**Comment by PVince81** [comment]
@owncloud-bot ok to test

**Comment by PVince81** [comment]
You might want to rebase on top of master in case this PR needs the styling fixes from https://github.com/owncloud/core/pull/13364 which was merged on master

**Comment by PVince81** [comment]
Works :+1: 

But I wonder about the private key which is stored in plain text inside mount.json.

Is it worth encrypting the private key like we did for passwords there ? CC @LukasReschke 
This could be done simply by renaming the field "private_key" to "password" or adding the field name here: https://github.com/owncloud/core/blob/master/apps/files_external/lib/config.php#L773 and here https://github.com/owncloud/core/blob/master/apps/files_external/lib/config.php#L790

**Comment by PVince81** [comment]
I'm merging this now as we need the AjaxController for other stuff.

The "private_key" encryption can be submit as a separate PR, if wanted.

Thanks a lot @rnicoll !

**Comment by rnicoll** [comment]
Private key encryption occurs at https://github.com/owncloud/core/pull/13190/files#diff-fc861a6589e68fe5eaca28c5e2f03296R23 based on the system's secret value. Please do let me know if there's any problems there.

**Comment by PVince81** [comment]
@rnicoll ah ok. I wasn't sure because when I looked at "mount.json" I saw the block "--- PRIVATE KEY BEGIN ---" or something. It looked like it was just the plain key.
If it's encrypted then it's fine.

**Pull request merged**